go/pools/resource_pool.go

```diff
@@ -131,14 +154,16 @@ func (rp *ResourcePool) closeIdleResources() {
 			return
 		}
 
-		if wrapper.resource != nil && idleTimeout > 0 && time.Until(wrapper.timeUsed.Add(idleTimeout)) < 0 {
-			wrapper.resource.Close()
-			wrapper.resource = nil
-			rp.idleClosed.Add(1)
-			rp.active.Add(-1)
-		}
+		func() {
+			defer func() { rp.resources <- wrapper }()
+
+			if wrapper.resource != nil && idleTimeout > 0 && time.Until(wrapper.timeUsed.Add(idleTimeout)) < 0 {
+				wrapper.resource.Close()
+				rp.idleClosed.Add(1)
+				rp.reopenResource(&wrapper)
```

Comment:
I wonder if a cleaner implementation that would obviate the need for the manually refactored `reopenResource` would be for the idle closer to actually just call `get` and `put` internally.

This would require actually keeping the `wait` parameter that you removed, but it seems like changing this to call `get` to obtain the resource, then closing it, then calling `put` to return it back means that we only would need to do the reopen in one place.

Reply:
OK -- what if instead we didn't reintroduce `wait`, but instead we just call in with a short context deadline.
In the usual case we will immediately get the connection since the whole thing is going to only grab what it thinks is `available`.

Then we use ErrTimeout to indicate that we've gone through all we need to, instead of the "stop early" case above.

Reply:
The main thing I don't like about this approach (both before and after your change) is that it duplicates a lot of the logic related to the stats and reopening the resource, etc.

If we use the existing get() / put() interface then `closeIdleResources` just operates like any other client to the pool -- it just grabs a connection to quickly close it, then rely on put() to reopen it and all the stats management would be done in get/put.

Reply:
(I actually wish I had done that originally tbh)

Reply:
Another related point which just came up in an internal discussion -- it would be nice to have vitess implement a clean mysql shutdown flow for the mysql protocol connections.

Currently our logs are filled up with things like:
`2019-05-28T16:11:52.895744-08:00 163 [Note] Aborted connection 163 to db: 'vt_byuser' user: 'vt_app_a' host: 'localhost' (Got an error reading communication packets)`

That's because Close() simply shuts down the underlying tcp socket.

I think instead we could bound the clean shutdown handshake by a relatively short context deadline (say 1-2 seconds) after which we summarily close the socket anyway.

That, to me, biases again for using the regular Get/Put interface for this.

Reply:
It turns out that we can't use `get` here because it's the wrapper that has the `timeUsed` metadata. We could look at changing `get` to return the wrapper instead, or an empty one if no resource was available. WDYT?

Reply:
Can we check this in and do the refactor later if we still want it?

Reply:
I think it'd be worthwhile to change `get` to return the possibly-nil-or-empty wrapper and thereby DRY this up and make the idle closer a bit cleaner.